src/main/java/com/google/devtools/build/lib/remote/SimpleBlobStoreActionCache.java

```diff
@@ -135,54 +125,30 @@ public void upload(
     }
 
     for (Map.Entry<Digest, Path> entry : manifest.getDigestToFile().entrySet()) {
-      try (InputStream in = entry.getValue().getInputStream()) {
-        uploadStream(entry.getKey(), in);
-      }
+      uploadFile(entry.getKey(), entry.getValue());
     }
 
     for (Map.Entry<Digest, Chunker> entry : manifest.getDigestToChunkers().entrySet()) {
-      uploadBlob(entry.getValue().next().getData().toByteArray(), entry.getKey());
-    }
-  }
-
-  public void uploadOutErr(ActionResult.Builder result, byte[] stdout, byte[] stderr)
-      throws IOException, InterruptedException {
-    if (stdout.length <= MAX_BLOB_SIZE_FOR_INLINE) {
-      result.setStdoutRaw(ByteString.copyFrom(stdout));
-    } else if (stdout.length > 0) {
-      result.setStdoutDigest(uploadBlob(stdout));
-    }
-    if (stderr.length <= MAX_BLOB_SIZE_FOR_INLINE) {
-      result.setStderrRaw(ByteString.copyFrom(stderr));
-    } else if (stderr.length > 0) {
-      result.setStderrDigest(uploadBlob(stderr));
+      uploadBlob(entry.getKey(), entry.getValue().next().getData());
     }
   }
 
-  public Digest uploadBlob(byte[] blob) throws IOException, InterruptedException {
-    return uploadBlob(blob, digestUtil.compute(blob));
-  }
-
-  private Digest uploadBlob(byte[] blob, Digest digest) throws IOException, InterruptedException {
-    try (InputStream in = new ByteArrayInputStream(blob)) {
-      return uploadStream(digest, in);
+  public void uploadFile(Digest digest, Path file) throws IOException, InterruptedException {
+    try {
+      blobStore.uploadFile(digest, file).get();
+    } catch (ExecutionException e) {
+      Throwables.propagateIfPossible(e.getCause(), IOException.class, InterruptedException.class);
+      throw new IOException(String.format("Uploading of file '%s' failed: %s", file.getPathString(), e.getCause()));
     }
   }
 
-  public Digest uploadStream(Digest digest, InputStream in)
-      throws IOException, InterruptedException {
-    final String hash = digest.getHash();
-
-    if (storedBlobs.putIfAbsent(hash, true) == null) {
```

Comment:
It seems that with this logic we tried to avoid uploading to BlobStore duplicates. Is it safe to just remove it now?

Reply:
I just moved the logic to `HttpBlobStore`. I believe that's where it belongs. For the `GrpcRemoteCache` the `ByteStreamUploader` also has its duplication logic. I can see good arguments for either having it in the AbstractRemoteCache or the protocol implementation, but I'd prefer to go with doing the de-duplication in the protocol implementation for simplicity.

We can re-evaluate once SimpleblobstoreActionCache is no more?

Reply:
ok, sounds good! Let's revisit it later :)